### PR TITLE
Nav block: handle overflowing items using wrap on Block selection

### PR DESCRIPTION
**Pull request by getdave**
An alternative attempt at https://github.com/WordPress/gutenberg/issues/18298.

See also https://github.com/WordPress/gutenberg/pull/18336#issuecomment-551534767

This approach avoids retaining the overflow scroll when the Block is in edit mode. Instead the items are allowed to wrap thereby improving the editing experience.

The major issue is that there is quite a big change in layout when the Block is selected.

## How has this been tested?
Manual testing.

## Screenshots 

Screen Capture on 2019-11-11 at 12-04-16

## Types of changes
Bug fix (non-breaking change which fixes an issue).

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

File: packages/block-library/src/navigation-menu/editor.scss [hunk 1]
```diff
@@ -2,6 +2,33 @@
...
+		}
+
+		.wp-block-navigation-menu .block-editor-block-list__layout {
+			display: flex;
+			flex-wrap: wrap;
+
+			.wp-block {
+				display: flex;
+				flex: 1; // grow to own content width only
+				// margin-top: $block-padding*2;
```
Comment (retrofox) [inline packages/block-library/src/navigation-menu/editor.scss]: are we going to keep these lines here?

File: packages/block-library/src/navigation-menu/edit.js [hunk 2]
```diff
@@ -45,6 +48,19 @@ function NavigationMenu( {
...
+	const navMenuRef = useRef();
+	const [ hasScrollX, setHasScrollX ] = useState( false );
+	const clientWidth = navMenuRef.current ? navMenuRef.current.clientWidth : 0;
+	const scrollWidth = navMenuRef.current ? navMenuRef.current.scrollWidth : 0;
+
+	useLayoutEffect( () => {
+		if ( scrollWidth > clientWidth ) {
+			setHasScrollX( true );
+		} else {
+			setHasScrollX( false );
```
Comment (retrofox) [inline packages/block-library/src/navigation-menu/edit.js]: minor suggestion
what do you think about 
```es6
useLayoutEffect( () => setHasScrollX( scrollWidth > clientWidth ), [ clientWidth, scrollWidth ] );
```

Feel free to ignore it 👍

File: packages/block-library/src/navigation-menu/edit.js [hunk 3]
```diff
@@ -79,6 +95,7 @@ function NavigationMenu( {
 		'wp-block-navigation-menu', {
 			'has-text-color': textColor.color,
 			'has-background-color': backgroundColor.color,
+			'has-scroll-x': hasScrollX,
```
Comment (retrofox) [inline packages/block-library/src/navigation-menu/edit.js]: It seems that the `has-scroll-x'` is not being used in any place of the CSS. Should we remove entirely the js code which handles it?

**Comment by jasmussen** [comment]
I think this also touches #18337.

File: packages/block-library/src/navigation-menu/editor.scss (hunk 1 above)
Comment (getdave) [reply packages/block-library/src/navigation-menu/editor.scss]: Yeh these are the key to the PR 😁

File: packages/block-library/src/navigation-menu/edit.js (hunk 3 above)
Comment (getdave) [reply packages/block-library/src/navigation-menu/edit.js]: We should if we decide this wrapping approach is valid.

File: packages/block-library/src/navigation-menu/edit.js (hunk 2 above)
Comment (getdave) [reply packages/block-library/src/navigation-menu/edit.js]: Yes that is valid.

**Comment by jasmussen** [comment]
Thank you for tackling this head-on. I'm seeing a few issues with scrollbars. This one probably shouldn't be present:

shift

There's also a shift.

There's also some sizing stuff going on:

typing

It is very possible and even likely that I'm missing some of the complexities of the navigation menu, but just to be absolutely sure it's not a failure of [my explanation](https://github.com/WordPress/gutenberg/pull/18336#issuecomment-552310979), I created #18431 as a PR that simply allows wrapping using `flex-wrap: wrap;`. It looks like this:

Screenshot 2019-11-11 at 14 52 46

It's simple, yes, but IMO it feels like a good and very basic _baseline_ that unblocks us, and then we can explore separate PRs that keep all menu items on a single line and handle overflow in a dropdown menu or hamburger or something in that vein. 

Did I miss anything obvious?

**Comment by draganescu** [comment]
To second @jasmussen why would we wrap only on edit and not all the time?

**Comment by shaunandrews** [comment]
A scrollbar seems very clunky. Perhaps we collapse overflowing menu items within a "more" button. This is a fairly common pattern across the web, and would closely match how I'd expect this to work on the frontend as well. Here's a quick GIF showing how this could look/work:

Navigation Overflow Menu

**Comment by draganescu** [comment]
@shaunandrews that is a creative approach! I like it :) I wonder what happens next:

- how do submenus work in that setup?
- how does the submenu item look with the label on?

**Comment by jasmussen** [comment]
> Perhaps we collapse overflowing menu items within a "more" button. This is a fairly common pattern across the web, and would closely match how I'd expect this to work on the frontend as well. 

Nice. It echoes [my thoughts](https://github.com/WordPress/gutenberg/pull/18336#issuecomment-552310979). But I wonder if the interface for _editing_ the menu should just let items wrap, as this 3 line PR does: #18431. Picture this — you select the block to edit it, and items wrap. You deselect it and overflowing items are moved into a dropdown menu.

The benefit of that approach is that we can merge #18431 as is to fix #18298, and then we can followup with a PR to enhance that further.

**Comment by getdave** [comment]
> The benefit of that approach is that we can merge #18431 as is to fix #18298, and then we can followup with a PR to enhance that further.

@jasmussen I'm confused as to how https://github.com/WordPress/gutenberg/pull/18431 improves upon this PR? Isn't it a duplicate?

This discussion feels like we need a decision. I feel this might be best to [happen on the Issue](https://github.com/WordPress/gutenberg/issues/18298) so it's more easily traceable. Do we agree?

**Comment by jasmussen** [comment]
It is a duplicate, yes, it's a different approach to accomplish wrapping. What I'm suggesting with followup improvements, is something along the lines of what Shaun has suggested with a dropdown menu. I don't necessarily think that one blocks anything.

> This discussion feels like we need a decision. I feel this might be best to happen on the Issue so it's more easily traceable. Do we agree?

Agree on keeping it on the issue, this was spread across three PRs so a bit hard to follow.

What needs deciding, though, scrollbar or wrap?

**Comment by getdave** [comment]
I've [moved the convo to the Issue](https://github.com/WordPress/gutenberg/issues/18298#issuecomment-553841060).

**Comment by getdave** [comment]
Closed in favour of Wrap solution. See Closing in favour of Wrap solution. See https://github.com/WordPress/gutenberg/issues/18298#issuecomment-553868002
